stdlib/public/core/AutoDiff.swift

```diff
@@ -495,6 +495,269 @@ public func gradient<T, U, V, R>(
+      fatalError()
+    }
+    return _ConcreteDerivativeBox(_baseDerivative - xBase)
+  }
+
+  // `Differentiable` requirements.
+
+  var _allDifferentiableVariables: _AnyDerivativeBox {
+    if _isOpaqueZero() {
+      return self
+    }
+    return _ConcreteDerivativeBox(_baseDerivative.allDifferentiableVariables)
+  }
+
+  func _moved(along direction: _AnyDerivativeBox) -> _AnyDerivativeBox {
+    if _isOpaqueZero() {
+      return direction
+    }
+    if direction._isOpaqueZero() {
+      return self
+    }
+    guard let directionBase =
+      direction._unboxed(to: T.TangentVector.self) else {
+      fatalError()
+    }
+    return _ConcreteDerivativeBox<T>(
+      _baseDerivative.moved(along: directionBase))
+  }
+
+  func _tangentVector(from cotangent: _AnyDerivativeBox) -> _AnyDerivativeBox {
+    if _isOpaqueZero() {
+      return type(of: cotangent)._cotangentZero
+        ._tangentVector(from: cotangent)
+    }
+    if cotangent._isOpaqueZero() {
+      return cotangent
+    }
+    guard let cotangentBase =
+      cotangent._unboxed(to: T.CotangentVector.self) else {
+      fatalError()
+    }
+    return _ConcreteDerivativeBox<T.TangentVector>(
+      _baseDerivative.tangentVector(from: cotangentBase))
+  }
+}
+
+/// A type-erased derivative value.
+///
+/// The `AnyDerivative` type forwards its operations to an arbitrary underlying
+/// base derivative value conforming to `Differentiable` and
+/// `AdditiveArithmetic`, hiding the specifics of the underlying value.
+public struct AnyDerivative : Differentiable & AdditiveArithmetic {
+  internal var _box: _AnyDerivativeBox
+
+  internal init(box: _AnyDerivativeBox) {
+    self._box = box
+  }
+
+  /// The underlying base value.
+  public var base: Any {
+    return _box._base
+  }
+
+  /// Creates a type-erased derivative from the given derivative.
+  public init<T>(_ base: T)
+    where T : Differentiable, T.TangentVector == T,
+          T.AllDifferentiableVariables == T,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          T.CotangentVector == T.CotangentVector.AllDifferentiableVariables
+  {
+    self._box = _ConcreteDerivativeBox<T>(base)
+  }
+
+  public typealias TangentVector = AnyDerivative
+  public typealias CotangentVector = AnyDerivative
+  public typealias AllDifferentiableVariables = AnyDerivative
+
+  // `Equatable` requirements (implied by `AdditiveArithmetic`).
+  public static func == (lhs: AnyDerivative, rhs: AnyDerivative) -> Bool {
+    return lhs._box._isEqual(to: rhs._box)
+  }
+  public static func != (lhs: AnyDerivative, rhs: AnyDerivative) -> Bool {
+    return lhs._box._isNotEqual(to: rhs._box)
+  }
+
+  // `AdditiveArithmetic` requirements.
+
+  /// Internal struct representing an opaque zero value.
+  @_fixed_layout
+  @usableFromInline
+  internal struct _OpaqueZero : Differentiable & AdditiveArithmetic {}
```

Comment:
This type will be seen by the user if they print `AnyDerivative.zero`. Maybe remove the underscore.

Reply:
Good point. Done.